### PR TITLE
Introduce Option type and remove "No" classes

### PR DESCRIPTION
**Pull request by jcollard**
There are many things I like about this approach.

However, I feel I've gone a little overboard throughout and the code looks like OCaml / Haskell code now. For a C# project, I think this is a bad thing. I would love your perspective @smabuk .

No pressure if you don't have the bandwidth to take a look but I always appreciate your feedback.

I'm currently leaning toward, "I'm going to go for it and see what I hate about it down the line."

### CODE REVIEW

**Comment by smabuk** [comment]
I think I agree with the C# Language team when they were considering this as an addition. It makes a lot of the code overly complex and barely readable, when the nullable reference types are basically saying the same thing without making your brain look through the Option type.

If necessary I see no reason why you couldn't just write extensions to handle nullable reference types the same way. Isn't that basically what Linq does with things like FirstOrDefault()

I went and rewatched Zorvan's videos on Option<T> and whilst I understand where he's coming from and the problems it solves I don't think I would go down that route in C#.

P.S. I checked out your code at the command line with a simple:
`gh pr checkout 46`

**Comment by jcollard** [comment]
I think the problems start to arise when mixing functional code with side effects. Option is great so long as the data is only be transformed. Once side effects are introduced, the code feels like it begins to unravel due to branching paths which are hard to follow within monadic function calls.

As always, I really appreciate your input and feedback!
